src/main/kotlin/org/jitsi/nlj/rtp/TransportCcEngine.kt

```diff
@@ -0,0 +1,151 @@
+import org.jitsi.utils.LRUCache
+import org.jitsi.utils.logging2.Logger
+
+/**
+ * Implements transport-cc functionality.
+ *
+ * See https://tools.ietf.org/html/draft-holmer-rmcat-transport-wide-cc-extensions-01
+ *
+ * @author Boris Grozev
+ * @author Julian Chukwu
+ * @author George Politis
+ */
+class TransportCcEngine(
+    private val bandwidthEstimator: BandwidthEstimator,
+    parentLogger: Logger,
+    private val clock: Clock = Clock.systemUTC()
+) : RtcpListener {
+
+    /**
+     * The [Logger] used by this instance for logging output.
+     */
+    private val logger: Logger = parentLogger.createChildLogger(javaClass.name)
+
+    /**
+     * Used to synchronize access to [.sentPacketDetails].
+     */
+    private val sentPacketsSyncRoot = Any()
+
+    /**
+     * The reference time of the remote clock. This is used to rebase the
+     * arrival times in the TCC packets to a meaningful time base (that of the
+     * sender). This is technically not necessary and it's done for convenience.
+     */
+    private var remoteReferenceTime: Instant = NEVER
+
+    /**
+     * Local time to map to the reference time of the remote clock. This is used
+     * to rebase the arrival times in the TCC packets to a meaningful time base
+     * (that of the sender). This is technically not necessary and it's done for
+     * convenience.
+     */
+    private var localReferenceTime: Instant = NEVER
+
+    /**
+     * Holds a key value pair of the packet sequence number and an object made
+     * up of the packet send time and the packet size.
+     */
+    private val sentPacketDetails = LRUCache<Int, PacketDetail>(MAX_OUTGOING_PACKETS_HISTORY)
+
+    /**
+     * Called when an RTP sender has a new round-trip time estimate.
+     */
+    fun onRttUpdate(rtt: Duration) {
+        val now = clock.instant()
+        bandwidthEstimator.onRttUpdate(now, rtt)
+    }
+
+    override fun rtcpPacketReceived(rtcpPacket: RtcpPacket, receivedTime: Long) {
+        if (rtcpPacket is RtcpFbTccPacket) {
+            tccReceived(rtcpPacket)
+        }
+    }
+
+    private fun tccReceived(tccPacket: RtcpFbTccPacket) {
+        val now = clock.instant()
+        var currArrivalTimestamp = Instant.ofEpochMilli(tccPacket.GetBaseTimeUs() / 1000)
+        if (remoteReferenceTime == NEVER) {
+            remoteReferenceTime = currArrivalTimestamp
+            localReferenceTime = now
+        }
+
+        for (packetReport in tccPacket) {
+            val packetDetail: PacketDetail?
+            val tccSeqNum = packetReport.seqNum
+            synchronized(sentPacketsSyncRoot) {
+                packetDetail = sentPacketDetails.remove(tccSeqNum)
+            }
+
+            if (packetDetail == null) {
+                if (packetReport is ReceivedPacketReport) {
+                    logger.warn("Couldn't find packet detail for $tccSeqNum.")
+                }
+                continue
+            }
+
+            if (!(packetReport is ReceivedPacketReport)) {
```

Comment:
should we just explicitly check for `packetReport is UnreceivedPacketReport` here?

Reply:
I've restructured this.

Any idea what to do if `packetReport` is neither `ReceivedPacketReport` or `UnreceivedPacketReport`?

Reply:
if we use a sealed class there we won't have to worry about that case (as it'll enforce those as the only possible options)